src/installer/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs

```diff
@@ -127,26 +132,40 @@ void RewriteAppHost(MemoryMappedFile mappedFile, MemoryMappedViewAccessor access
                         appHostSourceStream = new FileStream(appHostSourceFilePath, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 1);
                         memoryMappedFile = MemoryMappedFile.CreateFromFile(appHostSourceStream, null, 0, MemoryMappedFileAccess.Read, HandleInheritability.None, true);
                         memoryMappedViewAccessor = memoryMappedFile.CreateViewAccessor(0, 0, MemoryMappedFileAccess.CopyOnWrite);
-
                         // Get the size of the source app host to ensure that we don't write extra data to the destination.
                         // On Windows, the size of the view accessor is rounded up to the next page boundary.
-                        long sourceAppHostLength = appHostSourceStream.Length;
+                        var appHostLength = appHostSourceStream.Length;
 
                         // Transform the host file in-memory.
                         RewriteAppHost(memoryMappedFile, memoryMappedViewAccessor);
 
-                        // Save the transformed host.
-                        using (FileStream fileStream = new FileStream(appHostDestinationFilePath, FileMode.Create))
+                        MachObjectFile machObjectFile = null;
+                        if (!appHostIsPEImage)
                         {
-                            BinaryUtils.WriteToStream(memoryMappedViewAccessor, fileStream, sourceAppHostLength);
-
-                            // Remove the signature from MachO hosts.
-                            if (!appHostIsPEImage)
+                            if (enableMacOSCodeSign)
                             {
-                                MachOUtils.RemoveSignature(fileStream);
+                                string fileName = Path.GetFileName(appHostDestinationFilePath);
+                                machObjectFile = MachObjectFile.Create(memoryMappedViewAccessor, fileName);
+                                appHostLength = machObjectFile.CreateAdHocSignature(memoryMappedViewAccessor, fileName);
                             }
+                            else
+                            {
+                                if (MachObjectFile.TryRemoveCodesign(memoryMappedViewAccessor, out long? length))
+                                {
+                                    appHostLength = length.Value;
+                                }
+                            }
+                        }
 
-                            if (assemblyToCopyResourcesFrom != null && appHostIsPEImage)
+                        // Save the transformed host.
+                        using (FileStream fileStream = new FileStream(appHostDestinationFilePath, FileMode.Create, FileAccess.ReadWrite))
```

Comment:
Yeah, I felt it would be helpful to be consistent and explicit with all the different file-related IDisposable objects being opened, some with required access arguments and some without.